### DISCUSSION TITLE
apollo ProperySource 是怎么更新的？

### DISCUSSION THREAD

**Discussion by su-yh** [apolloconfig/apollo-java#14] [Q&A]
我看了下源代码，搭建了个基于spring boot 的简单环境。

我的理解是：

apollo client 更新配置主要是通过更新修改`PropertySource` 然后再遍历所有的`@Value` 注解并更新值，来达到实时变更配置属性的目的。

但是我有一个地方始终没有找到，就是apollo client 是在哪里对`PropertySource` 进行更新调用的。我看到apollo 加了一个`ApolloPropertySources` 属性源，并添加到优先级比较高的位置。但是他的值是在哪里进行更新的呢？

**Comment by nisiyong** [comment]
I think you could start debugging from the following class: 
1. `com.ctrip.framework.apollo.internals.RemoteConfigRepository`
2. `com.ctrip.framework.apollo.internals.RemoteConfigLongPollService`

It is very helpful to understand how the apollo java client works.

**Reply by su-yh** [reply]
这两个类我之前也已经跟踪调试过了，但是没找到我想要的答案。

我找到`com.ctrip.framework.apollo.internals.DefaultConfig` 似乎是它里面有监听器，但是我没看明白这里的数据变化是怎么使得spring 的`ConfigurableEnvironment`中的`PropertySource` 的数据发生变更 的。

我只看到它这里面的值的确发生的变化，但是没有看到触发的时机在哪里。

我跟踪到方法`com.ctrip.framework.apollo.spring.property.AutoUpdateConfigChangeListener#resolvePropertyValue()`
```java
  private Object resolvePropertyValue(SpringValue springValue) {
    // value will never be null, as @Value and @ApolloJsonValue will not allow that
    Object value = placeholderHelper
        .resolvePropertyValue(beanFactory, springValue.getBeanName(), springValue.getPlaceholder());

    ...
    ...
  }
```
这段代码`placeholderHelper.resolvePropertyValue()`执行之后就会得到最新的值，这个值就应该是从`ConfigurableEnvironment` 中的`PropertySource` 中得到。
所以我没看明白它是怎么被更新进`PropertySource`的。

我也看到代码中用到了spring boot 的`org.springframework.boot.context.config.ConfigData` ，它能够识别到`com.ctrip.framework.apollo.Config` 中的数据变更而自动更新到`PropertySource`吗？

**Reply by nobodyiam** [reply]
PropertySource 不需要更新，Spring 每次都是读取最新值的。

**Reply by su-yh** [reply]
我好像理出来了，嘿嘿。

apollo 在spring `ConfigurableEnvironment` 中添加了一个名为"ApolloPropertySources"的 `PropertySource`，类型为`CompositePropertySource`(spring 框架)。

在`CompositePropertySource` 中又添加了一个`PropertySource`，类型为`ConfigPropertySource`(apollo)。

`ConfigPropertySource` 是一个`PropertySource<Config>` 的一个实现类。

这一连串下来，在spring 中名为"ApolloPropertySources" 对应的`PropertySource` 的最终具体实现就是这个`Config` 接口，它的默认实现是`DefaultConfig`  。

所以只要`Config` 中对应的属性发生了变更也就是实际上"ApolloPropertySources" 的`PropertySource` 的值发生了变更。

因为：在调用`ConfigurableEnvironment` 父类中的方法: `getProperty(String)` 时，最终就会调用到`Config#getProperty(String, String)` 。

而apollo 的实现就是通过监听器的实现来修改`DefaultConfig` 中的值，同时也达到了 spring `ConfigurableEnvironment` 中的值变更的目的。

`DefaultConfig` 也实现了apollo 的监听器。监听着apollo 服务器数据变更事件。

`placeholderHelper.resolvePropertyValue()` 这里最终也就是调用了spring 的`PropertySourcesPlaceholderConfigurer` 提供的解析功能，获取到最新的值。


这个有些太绕了，主要是为什么中间还有一个`CompositePropertySource` ，如果直接是`ApolloPropertySources` 就没那么复杂 了。
